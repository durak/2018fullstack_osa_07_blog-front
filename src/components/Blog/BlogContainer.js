import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { blogLike, blogDestroy, blogAddComment } from '../../reducers/blogReducer'
import { notify } from '../../reducers/notificationReducer'

import Blog from './blogContainer/Blog'
import BlogComments from './blogContainer/BlogComments'
import BlogCommentForm from './blogContainer/BlogCommentForm'

export const BlogContainer = ({
  user,
  blogId,
  blogs,
  blogLike,
  blogDestroy,
  blogAddComment,
  notify }) => {

  const blog = blogs.find((b) => b.id === blogId)

  const handleDestroy = () => {
    if (window.confirm(`delete ${blog.title} by ${blog.author}?`)) {
      blogDestroy(blog)
      notify(`Blog ${blog.title} by ${blog.author} deleted`, 'message')
    }
  }

  const handleLike = () => {
    const msg = `you liked ${blog.title} by ${blog.author}`
    blogLike(blog)
    notify(msg, 'message')
  }

  const handleComment = (comment) => {
    const msg = `comment '${comment}' added to blog '${blog.title}'`
    blogAddComment(blog, comment)
    notify(msg, 'message')
  }


  return (
    <div>
      <Blog
        blog={blog}
        user={user}
        handleLike={handleLike}
        handleDestroy={handleDestroy}
      />
      <BlogComments blog={blog} />
      <BlogCommentForm handleComment={handleComment} />
    </div>
  )
}


const mapStateToProps = (state) => {
  return {
    blogs: state.blogs,
    user: state.loggedIn
  }
}

export default connect(
  mapStateToProps,
  { blogLike, blogDestroy, blogAddComment, notify }
)(BlogContainer)

BlogContainer.propTypes = {
  user: PropTypes.object.isRequired,
  blogId: PropTypes.string.isRequired,
  blogs: PropTypes.arrayOf(PropTypes.object).isRequired,
  blogLike: PropTypes.func.isRequired,
  blogAddComment: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired
}