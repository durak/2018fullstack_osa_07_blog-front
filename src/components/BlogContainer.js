import React from 'react'
import { connect } from 'react-redux'

import { blogLike, blogDestroy, blogAddComment } from '../reducers/blogReducer'
import { notify } from '../reducers/notificationReducer'

import Blog from './Blog'
import BlogComments from './BlogComments'
import BlogCommentForm from './BlogCommentForm'

const BlogContainer = ({
  user,
  blogId,
  blogs,
  blogLike,
  blogDestroy,
  blogAddComment,
  notify
}) => {

  console.log('blogs in BlogContainer', blogs)
  console.log('blogId in BlogContainer', blogId)
  console.log('user in BlogContainer', user)

  const blog = blogs.find((b) => b.id === blogId)

  const handleDestroy = () => {
    if (window.confirm(`delete ${blog.title} by ${blog.author}?`)) {
      blogDestroy(blog)
      notify(`Blog ${blog.title} by ${blog.author} deleted`, 'message')
    }
  }

  const handleLike = () => {
    const msg = `you liked ${blog.title} by ${blog.author}`
    blogLike(blog)
    notify(msg, 'message')
  }

  const handleComment = (comment) => {
    const msg = `comment '${comment}' added to blog '${blog.title}'`
    blogAddComment(blog, comment)
    notify(msg, 'message')
  }


  return (
    <div>
      <Blog
        blog={blog}
        user={user}
        handleLike={handleLike}
        handleDestroy={handleDestroy}
      />
      <BlogComments blog={blog} />
      <BlogCommentForm handleComment={handleComment} />
    </div>
  )
}


const mapStateToProps = (state) => {
  return {
    blogs: state.blogs,
    user: state.loggedIn
  }
}

export default connect(
  mapStateToProps,
  { blogLike, blogDestroy, blogAddComment, notify }
)(BlogContainer)