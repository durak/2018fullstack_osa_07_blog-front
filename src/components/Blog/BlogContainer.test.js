import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { BlogContainer } from './BlogContainer'



describe('<BlogContainer />', () => {

  const user = {
    username: 'testusername 1'
  }

  const blog1 = {
    id: '1',
    title: 'test blog 1',
    author: 'test author 1',
    url: 'http://testurl.com 1',
    likes: 99,
    user: { name: 'test user full name 1', username: 'testusername 1' },
    comments: [
      { _id: '1', comment: 'some comment content' },
      { _id: '2', comment: 'some other content' }
    ]
  }

  const blog2 = {
    id: '2',
    title: 'test blog 2',
    author: 'test author 2',
    url: 'http://testurl.com 2',
    likes: 99,
    user: { name: 'test user full name 2', username: 'testusername 2' },
    comments: [
      { _id: '1', comment: 'some comment content' },
      { _id: '2', comment: 'some other content' }
    ]
  }

  const mockAddComment = jest.fn()
  const mockBlogLike = jest.fn()
  const mockNotify = jest.fn()

  it('renderss correctly', () => {
    const tree = renderer.create(
      <BlogContainer
        user={user}
        blogId='1'
        blogs={[blog1, blog2]}
        blogLike={mockBlogLike}
        blogAddComment={mockAddComment}
        notify={mockNotify}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot()
  })

})