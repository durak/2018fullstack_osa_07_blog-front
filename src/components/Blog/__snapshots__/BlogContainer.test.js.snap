// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogContainer /> renderss correctly 1`] = `
<div>
  <div>
    <div
      className="ui fluid card"
      href={undefined}
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <h1
          className="header"
        >
          <span
            className="blogtitle"
          >
            test blog 1
          </span>
          <span
            className="right floated"
          >
            <button
              className="ui negative button blogDelete"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              style={
                Object {
                  "display": "",
                }
              }
              tabIndex={undefined}
            >
              delete
            </button>
          </span>
        </h1>
        <div
          className="meta"
        >
          <span>
            Author: 
            test author 1
          </span>
        </div>
        <div
          className="meta"
        >
          <span>
            Added by: 
            test user full name 1
          </span>
        </div>
        <div
          className="description"
        >
          url: 
          <a
            href="http://testurl.com 1"
            rel="noopener noreferrer"
            target="_blank"
          >
            http://testurl.com 1
          </a>
        </div>
      </div>
      <div
        className="extra content"
      >
        <span>
          <i
            aria-hidden="true"
            className="like icon blogVote"
            onClick={[Function]}
          />
          99
           votes
        </span>
        <span
          className="right floated"
        >
          <i
            aria-hidden="true"
            className="comments icon"
          />
          2
           comments
        </span>
      </div>
    </div>
  </div>
  <div
    className="ui comments"
  >
    <h3
      className="ui dividing header"
    >
      Comments
    </h3>
    <div
      className="comment"
    >
      <i
        aria-hidden="true"
        className="caret right icon avatar"
      />
      <div
        className="content"
      >
        <div
          className="text"
        >
          some comment content
        </div>
      </div>
    </div>
    <div
      className="comment"
    >
      <i
        aria-hidden="true"
        className="caret right icon avatar"
      />
      <div
        className="content"
      >
        <div
          className="text"
        >
          some other content
        </div>
      </div>
    </div>
  </div>
  <div>
    <form
      action={undefined}
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="inline field"
      >
        <div
          className="ui input"
        >
          <input
            disabled={undefined}
            name="comment"
            onChange={[Function]}
            tabIndex={undefined}
            type="text"
          />
        </div>
        <label>
          <output
            name="valid"
            type="text"
          />
        </label>
      </div>
      <button
        className="ui button"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
        type="submit"
      >
        add comment
      </button>
    </form>
  </div>
</div>
`;
